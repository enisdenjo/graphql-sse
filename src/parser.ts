/**
 *
 * parser
 *
 */

import {
  StreamMessage,
  validateStreamEvent,
  parseStreamData,
  StreamEvent,
} from './common';

enum ControlChars {
  NewLine = 10,
  CarriageReturn = 13,
  Space = 32,
  Colon = 58,
}

/**
 * HTTP response chunk parser for graphql-sse's event stream messages.
 *
 * Reference: https://github.com/Azure/fetch-event-source/blob/main/src/parse.ts
 *
 * @private
 */
export function createParser<ForID extends boolean>(): (
  chunk: Uint8Array,
) => StreamMessage<ForID, StreamEvent>[] | void {
  let buffer: Uint8Array | undefined;
  let position: number; // current read position
  let fieldLength: number; // length of the `field` portion of the line
  let discardTrailingNewline = false;
  let message = { event: '', data: '' };
  let pending: StreamMessage<ForID, StreamEvent>[] = [];
  const decoder = new TextDecoder();

  return function parse(chunk) {
    if (buffer === undefined) {
      buffer = chunk;
      position = 0;
      fieldLength = -1;
    } else {
      const concat = new Uint8Array(buffer.length + chunk.length);
      concat.set(buffer);
      concat.set(chunk, buffer.length);
      buffer = concat;
    }

    const bufLength = buffer.length;
    let lineStart = 0; // index where the current line starts
    while (position < bufLength) {
      if (discardTrailingNewline) {
        if (buffer[position] === ControlChars.NewLine) {
          lineStart = ++position; // skip to next char
        }
        discardTrailingNewline = false;
      }

      // look forward until the end of line
      let lineEnd = -1; // index of the \r or \n char
      for (; position < bufLength && lineEnd === -1; ++position) {
        switch (buffer[position]) {
          case ControlChars.Colon:
            if (fieldLength === -1) {
              // first colon in line
              fieldLength = position - lineStart;
            }
            break;
          // \r case below should fallthrough to \n:
          case ControlChars.CarriageReturn:
            discardTrailingNewline = true;
          // eslint-disable-next-line no-fallthrough
          case ControlChars.NewLine:
            lineEnd = position;
            break;
        }
      }

      if (lineEnd === -1) {
        // end of the buffer but the line hasn't ended
        break;
      } else if (lineStart === lineEnd) {
        // empty line denotes end of incoming message
        if (message.event || message.data) {
          // NOT a server ping (":\n\n")
          if (!message.event) throw new Error('Missing message event');
          const event = validateStreamEvent(message.event);
          const data = parseStreamData<ForID, StreamEvent>(event, message.data);
          pending.push({
            event,
            data,
          });
          message = { event: '', data: '' };
        }
      } else if (fieldLength > 0) {
        // end of line indicates message
        const line = buffer.subarray(lineStart, lineEnd);

        // exclude comments and lines with no values
        // line is of format "<field>:<value>" or "<field>: <value>"
        // https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation
        const field = decoder.decode(line.subarray(0, fieldLength));
        const valueOffset =
          fieldLength + (line[fieldLength + 1] === ControlChars.Space ? 2 : 1);
        const value = decoder.decode(line.subarray(valueOffset));

        switch (field) {
          case 'event':
            message.event = value;
            break;
          case 'data':
            // append the new value if the message has data
            message.data = message.data ? message.data + '\n' + value : value;
            break;
        }
      }

      // next line
      lineStart = position;
      fieldLength = -1;
    }

    if (lineStart === bufLength) {
      // finished reading
      buffer = undefined;
      const messages = [...pending];
      pending = [];
      return messages;
    } else if (lineStart !== 0) {
      // create a new view into buffer beginning at lineStart so we don't
      // need to copy over the previous lines when we get the new chunk
      buffer = buffer.subarray(lineStart);
      position -= lineStart;
    }
  };
}
