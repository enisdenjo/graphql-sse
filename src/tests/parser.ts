import { createParser } from '../parser';

const encoder = new TextEncoder();

it('should parse whole message', () => {
  const parse = createParser();

  // with space
  expect(
    parse(encoder.encode('id: 1\nevent: value\ndata: { "iAm": "data" }\n\n')),
  ).toMatchSnapshot();

  // no space
  expect(
    parse(encoder.encode('id:1\nevent:value\ndata:{ "iAm": "data" }\n\n')),
  ).toMatchSnapshot();

  // no data or id
  expect(parse(encoder.encode('event: done\n\n'))).toMatchSnapshot();
});

it('should parse chunked message', () => {
  const parse = createParser();

  parse(encoder.encode('id: 1\neven'));
  parse(encoder.encode(''));
  parse(encoder.encode('t: val'));
  parse(encoder.encode('ue\nda'));
  parse(encoder.encode('ta:{'));
  parse(encoder.encode(''));
  parse(encoder.encode(' "iAm'));
  const msg = parse(encoder.encode('": "data" }\n\n'));

  expect(msg).toMatchSnapshot();
});

it('should parse message whose lines are separated by \\r\\n', () => {
  const parse = createParser();

  const msg = parse(
    encoder.encode('id: 1\r\nevent: value\r\ndata: { "iAm": "data" }\r\n\r\n'),
  );

  expect(msg).toMatchSnapshot();
});

it('should ignore comments', () => {
  const parse = createParser();

  const msg = parse(
    encoder.encode(': I am comment\nevent: value\ndata: { "iAm": "data" }\n\n'),
  );

  expect(msg).toMatchSnapshot();
});

it('should accept valid events only', () => {
  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: complete\ndata: {}\n\n'));
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: next\ndata: {}\n\n'));
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: value\ndata: {}\n\n'));
  }).not.toThrow();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: done\ndata: {}\n\n'));
  }).not.toThrow();
});

it('should ignore server pings', () => {
  const parse = createParser();

  expect(parse(encoder.encode(':\n\n'))).toBeUndefined();
});

it('should parse multiple messages from one chunk', () => {
  const parse = createParser();

  expect(
    parse(
      encoder.encode(
        'id: 1\nevent: value\ndata: {}\n\n' +
          'id: 2\nevent: value\ndata: { "no": "data" }\n\n',
      ),
    ),
  ).toMatchSnapshot();

  expect(
    parse(
      encoder.encode(
        'id: 3\nevent: value\ndata: { "almost": "done" }\n\n' +
          'id: 4\nevent: done\ndata: {}\n\n',
      ),
    ),
  ).toMatchSnapshot();
});
