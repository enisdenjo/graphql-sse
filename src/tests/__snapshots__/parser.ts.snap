// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept valid events only 1`] = `"Invalid stream event \\"done\\""`;

exports[`should accept valid events only 2`] = `"Invalid stream event \\"value\\""`;

exports[`should ignore comments 1`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse chunked message 1`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse message whose lines are separated by \\r\\n 1`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse message with prepended ping 1`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse multiple messages from one chunk 1`] = `
Array [
  Object {
    "data": Object {},
    "event": "next",
  },
  Object {
    "data": Object {
      "no": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse multiple messages from one chunk 2`] = `
Array [
  Object {
    "data": Object {
      "almost": "done",
    },
    "event": "next",
  },
  Object {
    "data": Object {},
    "event": "complete",
  },
]
`;

exports[`should parse whole message 1`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse whole message 2`] = `
Array [
  Object {
    "data": Object {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse whole message 3`] = `
Array [
  Object {
    "data": null,
    "event": "complete",
  },
]
`;
